### PR TITLE
rough cut C++ emit for value IR

### PR DESCRIPTION
**Pull request by cseed**
- first cut C++ value IR compiler
 - followed existing design, e.g. cxx.EmitTriplet, but carry physical type,
 - added cxx.Compile, compiles a function takes and returns a single non-missing tuple,
 - throws CXXUnsupportedOperation if it can't compile,
 - integrate with assertEvalsTo

### CODE REVIEW

**Comment by cseed** [comment]
Added support for a few more nodes, including {Insert, Select}Fields and Array{Range, Map, Filter} (deforested).  I'm going to do ordering next.  That should give us enough interesting stuff to play with when simple Table pipeline start working.

File: hail/src/test/scala/is/hail/expr/ir/ArrayFunctionsSuite.scala [hunk 1]
```diff
@@ -53,11 +53,12 @@ class ArrayFunctionsSuite extends TestNGSuite {
 
   @Test def mean() {
     assertEvalsTo(invoke("mean", IRArray(3, 7)), 5.0)
+    /*
```
Comment (catoverdrive) [inline hail/src/test/scala/is/hail/expr/ir/ArrayFunctionsSuite.scala]: did you mean to comment this out?

**Comment by catoverdrive** [comment]
@cseed do you have tests for this? I don't see any right now.

**Comment by catoverdrive** [comment]
oh wait, I see. they got bundled into `assertEvalsTo`

**Comment by cseed** [comment]
> they got bundled into assertEvalsTo

Yes, and assertEvalSame.  Quite a bit of stuff is running through that, and it is definitely getting tested because I had to squash a bunch of bugs.

File: hail/src/test/scala/is/hail/expr/ir/ArrayFunctionsSuite.scala (hunk 1 above)
Comment (cseed) [reply hail/src/test/scala/is/hail/expr/ir/ArrayFunctionsSuite.scala]: No, was debugging something.  Fixed.

File: hail/src/main/scala/is/hail/cxx/Emit.scala [hunk 2]
```diff
@@ -0,0 +1,532 @@
...
+
+      case ir.NA(t) =>
+        triplet("", "true", typeDefaultValue(pType))
+
+      case ir.IsNA(v) =>
+        val t = emit(v)
+        triplet(t.setup, "false", t.m)
+
+      case ir.If(cond, cnsq, altr) =>
+        // FIXME
```
[the remark this answers is not in the document]
Comment (catoverdrive) [reply hail/src/main/scala/is/hail/cxx/Emit.scala]: (can you add a note about what needs fixing?)

File: hail/src/test/scala/is/hail/expr/ir/IRSuite.scala [hunk 3]
```diff
@@ -3,17 +3,20 @@ package is.hail.expr.ir
 import is.hail.SparkSuite
 import is.hail.expr.types._
 import is.hail.TestUtils._
+import is.hail.annotations.{Region, RegionValue, RegionValueBuilder, SafeRow}
```
[the remark this answers is not in the document]
Comment (cseed) [reply hail/src/test/scala/is/hail/expr/ir/IRSuite.scala]: Yes.  I have the habit of deleting unused ones when I notice them.

File: hail/src/main/scala/is/hail/cxx/Emit.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (cseed) [reply hail/src/main/scala/is/hail/cxx/Emit.scala]: I had already fixed it.  Removed.

File: hail/src/main/scala/is/hail/expr/types/physical/PContainer.scala [hunk 4]
```diff
@@ -222,4 +223,58 @@ abstract class PContainer extends PType {
...
+      s"(($len + 7) >> 3)"
+  }
+
+  def cxxContentsByteSize(len: cxx.Code): cxx.Code = {
+    s"${ cxxElementsOffset(len) } + $len * ${ elementByteSize }"
+  }
+
+  def cxxElementsOffset(len: cxx.Code): cxx.Code = {
+    if (elementType.required)
+      s"round_up_alignment(4, ${ elementType.alignment })"
```
[the remark this answers is not in the document]
Comment (cseed) [reply hail/src/main/scala/is/hail/expr/types/physical/PContainer.scala]: Small, obviously, but less C++ code to compile is probably a good thing.

File: hail/src/main/scala/is/hail/expr/types/physical/PContainer.scala [hunk 5]
```diff
@@ -222,4 +223,58 @@ abstract class PContainer extends PType {
...
+${ __i.define }
+if (${ a.m } || ${ i.m })
+  $m = true;
+else {
+  ${ __a } = ${ a.v };
+  ${ __i } = ${ i.v };
+  $m = ${ cxxIsElementMissing(__a.toString, __i.toString) };
+}
+""", m.toString,
+      cxx.loadIRIntermediate(elementType, cxxElementOffset(__a.toString, __i.toString)))
```
[the remark this answers is not in the document]
Comment (cseed) [reply hail/src/main/scala/is/hail/expr/types/physical/PContainer.scala]: I changed the code to create the analogous error message in NATIVE_STATUS and return.  Same for the `abort()`s.

File: hail/src/main/scala/is/hail/cxx/Emit.scala [hunk 6]
```diff
@@ -0,0 +1,532 @@
...
+        pStruct.fields.foreach { f =>
+          fieldsMap.get(f.name) match {
+            case Some(fx) =>
+              val fxt = emit(fx)
+              sb.add(fxt)
+            case None =>
+              val fieldIdx = oldPStruct.fieldIdx(f.name)
+              sb.add(
+                EmitTriplet(f.typ, "",
+                  oldPStruct.cxxIsFieldMissing(ov.toString, f.index),
```
[the remark this answers is not in the document]
Comment (cseed) [reply hail/src/main/scala/is/hail/cxx/Emit.scala]: You're 100% right.

File: hail/src/main/scala/is/hail/cxx/Emit.scala [hunk 7]
```diff
@@ -0,0 +1,532 @@
...
+"""
+        }
+
+        triplet(sb.result().mkString, "false", sab.end())
+
+      case x@ir.ApplyIR(_, _, _) =>
+        // FIXME small only
+        emit(x.explicitNode)
+
+      case ir.In(i, _) =>
```
[the remark this answers is not in the document]
Comment (cseed) [reply hail/src/main/scala/is/hail/cxx/Emit.scala]: Wasn't my intention.  I'm being a bit lazy here.  But I expect the C++ code generator, unlike the Java code generator, not be used to generate stand alone functions (except in tests with, as you say, restricted expectations?), but instead be called as part of the codegen for a larger pipeline.

I also think we can just delete `In` in favor of variable references, but that might require some work on the JVM side so we might be better of waiting until we just delete that to clean this up.

So I'm not quite sure what to do.  I say we build it up as needed and leave it restricted for now noting we might have to come back to this spot.

**Comment by cseed** [comment]
I had rebuilt them from debugging, but checked them in by accident.  Reverted.

**Pull request merged**